docs: Add charm architecture documentation (#19)
canonical/tmate-ssh-server-operator: docs/explanation/charm-architecture.md

    Container(tmatessh, "tmate SSH Server", "", "Provides sharing a remote session")
 }
Container_Boundary(c2, "tmate Client Charm") {
    Container(clientcharm, "tmate Client Charm", "", "Charm managing tmate client workload.")
 }

    System_Ext(tmateclient, "tmate Client", "Client that shares their terminal session")
    System_Ext(collaborator, "Collaborator", "Another client who wants access to the terminal")

    Rel(tmatessh, clientcharm, "provides connection details")
    Rel(clientcharm, tmateclient, "init client with connection details")

    Rel(tmateclient, tmatessh, "init session via SSH")
    Rel(collaborator, tmatessh, "join session via SSH")

    UpdateRelStyle(tmatessh, clientcharm, $offsetY="30", $offsetX="-70")
    UpdateRelStyle(clientcharm, tmateclient, $offsetY="80", $offsetX="150")
    UpdateRelStyle(tmateclient, tmatessh, $offsetY="-30", $offsetX="10")
    UpdateRelStyle(collaborator, tmatessh, $offsetY="-30", $offsetX="80")
```

### Why is this a machine charm?

The OCI image for tmate-ssh-server requires certain kernel capabilities to run ([SYS_ADMIN](https://github.com/tmate-io/tmate-ssh-server/issues/73#issuecomment-762756632)).

> At design time, due to the limitations of a Juju kubernetes charm, the container was unable to receive the capability.

Reviewer: ```suggestion
At design time, due to the limitations of a Juju Kubernetes charm, the container was unable to receive the capability.
```

> Furthermore, providing SSH ingress for machine charms is easier to achieve than for k8s charms. Hence, it was decided to make this a machine charm. 

Reviewer: ```suggestion
Furthermore, providing SSH ingress for machine charms is easier to achieve than for K8s charms. Hence, it was decided to make tmate-ssh-server a machine charm. 
```

# Charm architecture

The tmate-ssh-server-operator operates [tmate-ssh-server](https://github.com/tmate-io/tmate-ssh-server),
which is the server side of [tmate](https://tmate.io/), an open-source terminal multiplexer
that provides instant terminal sharing capabilities.
The tmate-ssh-server application is built from source as an OCI image and runs as a systemd service on the machine.

The charm provides the necessary connection details for a tmate client to connect to the tmate-ssh-server service
via integration data.

```mermaid
C4Container

> title Container diagram for Tmate SSH Server

Reviewer: For my understanding, why is it referred to as "tmate" in the text but "Tmate" in the diagram? I feel like we should stick to one capitalization choice for the entire document.

Author: The idea was to use title case here.

Reviewer: Thanks for the context. I still think that we should stick to "tmate" -- thanks for updating the diagram!

 Container_Boundary(c1, "Tmate SSH Server") {
    Container(tmatessh, "Tmate SSH Server", "", "Provides sharing a remote session")
 }
Container_Boundary(c2, "Tmate Client Charm") {
    Container(clientcharm, "Tmate Client Charm", "", "Charm managing Tmate Client workload.")
 }

>     System_Ext(tmateclient, "Tmate Client", "Client that shares his terminal session")

Reviewer: ```suggestion
    System_Ext(tmateclient, "Tmate Client", "Client that shares their terminal session")
```
Assuming that the longer word doesn't disrupt the aesthetics, I think it would be good to be gender neutral :)

    System_Ext(collaborator, "Collaborator", "Another client who wants access to the terminal")

    Rel(tmatessh, clientcharm, "provides connection details")
    Rel(clientcharm, tmateclient, "init client with connection details")

    Rel(tmateclient, tmatessh, "init session via ssh")
    Rel(collaborator, tmatessh, "join session via ssh")

    UpdateRelStyle(tmatessh, clientcharm, $offsetY="30", $offsetX="-70")
    UpdateRelStyle(clientcharm, tmateclient, $offsetY="80", $offsetX="150")
    UpdateRelStyle(tmateclient, tmatessh, $offsetY="-30", $offsetX="10")
    UpdateRelStyle(collaborator, tmatessh, $offsetY="-30", $offsetX="80")
```

## OCI images

> We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build OCI Images for tmate-ssh-server. 

Reviewer: ```suggestion
We use [Rockcraft](https://canonical-rockcraft.readthedocs-hosted.com/en/latest/) to build OCI images for tmate-ssh-server. 
```

The images are defined in [tmate-ssh-server_rock](https://github.com/canonical/tmate-ssh-server-operator/tree/main/tmate-ssh-server_rock).

> They are published to the [Github Container registry](https://github.com/canonical/tmate-ssh-server-operator/pkgs/container/tmate-ssh-server).

Reviewer 2: I think it is published to the github container registry, but is pushed to charmhub!
